### PR TITLE
[Enhance] Support modifying palette through configuration

### PR DESCRIPTION
**Pull request by hhaAndroid**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Resolved: https://github.com/open-mmlab/mmdetection/issues/9309

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Review by RangiLyu — COMMENTED** [review]
UT failed

File: mmdet/apis/inference.py [hunk 1]
```diff
@@ -65,16 +65,31 @@ def init_detector(
...
             warnings.warn(
                 'dataset_meta or class names are not saved in the '
                 'checkpoint\'s meta data, use COCO classes by default.')
-            model.dataset_meta = {
-                'CLASSES': get_classes('coco'),
-                'PALETTE': palette
-            }
+            model.dataset_meta = {'CLASSES': get_classes('coco')}
+
+    # Priority: config -> args.palette -> checkpoint
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply mmdet/apis/inference.py:76]: makes sense

File: mmdet/apis/inference.py [hunk 2]
```diff
@@ -65,16 +65,28 @@ def init_detector(
...
+        metainfo = config.test_dataloader.dataset.get('metainfo', {})
+        cfg_palette = metainfo.get('PALETTE', None)
+        if cfg_palette is not None:
+            model.dataset_meta['PALETTE'] = cfg_palette
+        else:
+            if 'PALETTE' not in model.dataset_meta:
+                warnings.warn(
+                    'PALETTE does not exist, coco is used by default. '
+                    'You can also set the palette to customize.')
+                model.dataset_meta['PALETTE'] = 'coco'
```
Comment (RangiLyu) [inline mmdet/apis/inference.py:89]: If there is no 'PALETTE', should the color be randomly set? If set it to coco by default, will it cause some error if the length of PALETTE smaller than the num_classes of the model?

Comment (hhaAndroid) [reply mmdet/apis/inference.py:89]: That's right, but this one is kept consistent with the previous implementation.

**Review by RangiLyu — APPROVED** [review]

**Review by RangiLyu — CHANGES_REQUESTED** [review]
Need to refactor CLASSES and PALETTE to lowercase

**Review by RangiLyu — APPROVED** [review]

File: mmdet/apis/inference.py [hunk 3]
```diff
@@ -54,31 +54,50 @@ def init_detector(
...
     if checkpoint is not None:
         checkpoint = load_checkpoint(model, checkpoint, map_location='cpu')
         # Weights converted from elsewhere may not have meta fields.
         checkpoint_meta = checkpoint.get('meta', {})
+
         # save the dataset_meta in the model for convenience
         if 'dataset_meta' in checkpoint_meta:
             # mmdet 3.x
-            model.dataset_meta = checkpoint_meta['dataset_meta']
+            dataset_meta = checkpoint_meta['dataset_meta']
```
Comment (ZwwWayne) [inline mmdet/apis/inference.py:66]: shall we directly use `model.dataset_meta=copy.deepcopy(checkpoint_meta['dataset_meta'])`

Comment (hhaAndroid) [reply mmdet/apis/inference.py:66]: There should be no need for deepcopy.

**Review by RangiLyu — APPROVED** [review]

**Pull request merged**